common/meta/api/src/schema_api_impl.rs

```diff
@@ -196,6 +263,101 @@ impl<KV: KVApi> SchemaApi for KV {
+
+            if res.is_ok() {
+                return Err(MetaError::AppError(AppError::UndropDbAlreadyExists(
+                    UndropDbAlreadyExists::new(&name_key.db_name),
+                )));
+            }
+
+            // get db id list from _fd_db_id_list/<tenant>/<db_name>
+            let dbid_idlist = DbIdListKey {
+                tenant: name_key.tenant.clone(),
+                db_name: name_key.db_name.clone(),
+            };
+            let (db_id_list_seq, db_id_list_opt): (_, Option<DbIdList>) =
+                get_struct_value(self, &dbid_idlist).await?;
+
+            let mut db_id_list = if db_id_list_seq == 0 {
+                return Err(MetaError::AppError(AppError::UndropDbHasNoHistory(
+                    UndropDbHasNoHistory::new(&name_key.db_name),
+                )));
+            } else {
+                match db_id_list_opt {
+                    Some(list) => list,
+                    None => {
+                        return Err(MetaError::AppError(AppError::UndropDbHasNoHistory(
+                            UndropDbHasNoHistory::new(&name_key.db_name),
+                        )))
+                    }
+                }
+            };
+
+            // Return error if there is no db id history.
+            let db_id = match db_id_list.last() {
+                Some(db_id) => *db_id,
+                None => {
+                    return Err(MetaError::AppError(AppError::UndropDbHasNoHistory(
+                        UndropDbHasNoHistory::new(&name_key.db_name),
+                    )))
+                }
+            };
+
+            // get tb_meta of the last db id
+            let dbid = DatabaseId { db_id };
+            let (db_meta_seq, db_meta): (_, Option<DatabaseMeta>) =
+                get_struct_value(self, &dbid).await?;
+
+            tracing::debug!(db_id, name_key = debug(&name_key), "undrop_database");
+
+            {
+                // reset drop on time
+                let mut db_meta = db_meta.unwrap();
+                // undrop a table with no drop time
+                if db_meta.drop_on.is_none() {
+                    return Err(MetaError::AppError(AppError::UndropDbWithNoDropTime(
+                        UndropDbWithNoDropTime::new(&name_key.db_name),
+                    )));
+                }
+                db_meta.drop_on = None;
+
+                let txn_req = TxnRequest {
+                    condition: vec![
+                        txn_cond_seq(name_key, Eq, 0)?,
+                        txn_cond_seq(&dbid_idlist, Eq, db_id_list_seq)?,
+                        txn_cond_seq(&dbid, Eq, db_meta_seq)?,
+                    ],
+                    if_then: vec![
+                        txn_op_put(name_key, serialize_id(db_id)?)?, // (tenant, db_name) -> db_id
+                        txn_op_put(&dbid, serialize_struct(&db_meta)?)?, // (db_id) -> db_meta
+                    ],
+                    else_then: vec![],
```

Comment:
shall we check the status of `name_key` in the `else` branch? so that if name conflict detected, just break out the loop?

Reply:
Oh, pls ignore it, I miss-interpreted the code